lib/fs-writable.js

```diff
@@ -0,0 +1,151 @@
+// Copyright Node.js contributors. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+'use strict';
+
+const Writable = require('./writable');
+const fs = require('fs');
+const {
+  ERR_INVALID_ARG_TYPE,
+  ERR_OUT_OF_RANGE
+} = require('./errors');
+
+const closeFsStream = (stream, cb, err) => {
+  fs.close(stream.fd, (er) => {
+    er = er || err;
+    cb(er);
+    stream.closed = true;
+    if (!er) stream.emit('close');
+  });
+};
+
+class WriteStream extends Writable {
+  constructor(path, {
+    fd = null,
+    start,
+    flags = 'w',
+    mode = 0o666,
+    encoding,
+    autoClose = true,
+    bufferSize
+  } = {}) {
+    super({ bufferSize });
+
+    this.path = path;
+    this.fd = fd;
+    this.flags = flags;
+    this.mode = mode;
+
+    this.start = start;
+    this.autoClose = autoClose;
+
+    this.closed = false;
+    this.pos = undefined;
+    this.bytesWritten = 0;
+
+    if (this.start !== undefined) {
+      if (typeof this.start !== 'number') {
+        throw new ERR_INVALID_ARG_TYPE('start', 'number', this.start);
+      }
+      if (this.start < 0) {
+        const errVal = `{start: ${this.start}}`;
+        throw new ERR_OUT_OF_RANGE('start', '>= 0', errVal);
+      }
+
+      this.pos = this.start;
+    }
+
+    if (encoding) this.setDefaultEncoding(encoding);
+
+    if (typeof fd !== 'number') this.open();
+  }
+
+  open() {
+    fs.open(this.path, this.flags, this.mode, (er, fd) => {
+      if (er) {
+        if (this.autoClose) this.destroy();
+        this.emit('error', er);
+        return;
+      }
+
+      this.fd = fd;
+      this.emit('open', fd);
+      this.emit('ready');
```

Comment:
I see no reason behind this event because you can replace all its usages with event `close` AFAIU.

Reply:
@nechaido I think you meant `open`?